File: in_class_day03_cpp_solutions/src/emergency_stop.cpp
#include <cstdio>
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "neato2_interfaces/msg/bump.hpp"
#include <atomic>
#include <chrono>

/// The Node that encapsulates our main functionality 
class EmergencyStop : public rclcpp::Node {
  public:
    EmergencyStop()
    : Node("estop")
    {
      publisher = this->create_publisher<geometry_msgs::msg::Twist>("cmd_vel", 10);
      bumper_subscription = this->create_subscription<neato2_interfaces::msg::Bump>(
          "bump",
          10,
          std::bind(
              &EmergencyStop::handle_bump,
              this,
              std::placeholders::_1));
      timer = rclcpp::create_timer(this,
                                   this->get_clock(),
                                   rclcpp::Duration::from_seconds(0.1),
                                   std::bind(&EmergencyStop::run_loop, this));
    }

    void handle_bump(const neato2_interfaces::msg::Bump::SharedPtr msg)
    {
      bumper_active = (msg->left_front == 1 ||
                       msg->right_front == 1 ||
                       msg->left_side == 1 ||
                       msg->right_side == 1);
    }

    void run_loop()
    {
      geometry_msgs::msg::Twist msg;
      if (bumper_active) {
        msg.linear.x = 0.0;
      } else {
        msg.linear.x = 0.1;
      }
      publisher->publish(msg);
    }

  private:
    rclcpp::Subscription<neato2_interfaces::msg::Bump>::SharedPtr bumper_subscription;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher;
    rclcpp::TimerBase::SharedPtr timer;
    std::atomic<bool> bumper_active;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<EmergencyStop>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: in_class_day03/in_class_day03/emergency_stop.py
import rclpy
from rclpy.node import Node
from neato2_interfaces.msg import Bump
from geometry_msgs.msg import Twist


class EmergencyStopNode(Node):
    """This node stops the the motors when bump is triggered"""

    def __init__(self):
        """Initializes the class."""
        super().__init__("emergencency_stop_node")
        self.create_timer(0.1, self.run_loop)  # timer runs the loop
        self.create_subscription(
            Bump, "bump", self.process_bump, 10
        )  # continually broadcasts messages
        self.vel_pub = self.create_publisher(Twist, "cmd_vel", 10)  #
        self.bumper_active = False

        # create a timer that runs the robot motors
        # create a state that stores the bumper information
        # create a subscriber to the bumper information
        # create a subscriber to the bump sensor
        # create a publisher for the motors

    def process_bump(self, msg):
        """callback from handling a bump sensor input"""
        # process message
        # determine if a bump has happened
        # set the bumper state variable
        self.bumper_active = (
            msg.left_front == 1
            or msg.left_side == 1
            or msg.right_front == 1
            or msg.right_side == 1
        )

    def run_loop(self):
        """Event loop drives the motors based on the bumper"""
        msg = Twist()
        msg.linear.x = 0.1 if not self.bumper_active else 0.0
        self.vel_pub.publish(msg)
        # command the robot to move
        # if a bump is sensed, stop robot motion


def main(args=None):
    """Initialization, Spin, Shutdown"""
    rclpy.init(args=args)
    node = EmergencyStopNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
